### PR TITLE
Do not sample numpy array when calculate brightness range

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues
hides #5608 and #6183 

# Description

This PR hides the problem of over-saturated images by always calculating min and max on images that are numpy arrays. 

The data from #6183 has size 1.411 x 1e7. 
The data from #5608 has size 3.41 x 1e7.

So workaround may be to limit the size of the numpy array to 1e8 for calculating over the whole image.

### CODE REVIEW

**Comment by jni** [comment]
@Czaki I find the implementation hard to follow, with too much special-casing. I think we can have a single n-dimensional sampler, something like this:

```python
def sample_data_chunks(data, sample_size, chunks=3):
    ndim = data.ndim
    chunk_size_1d = sample_size ** (1/ndim) / chunks**ndim
    margin = chunk_size_1d // 2
    midpoints = tuple(s // 2 for s in data.shape)
    slice_per_dim = [
        (slice(None, chunk_size_1d),  # start of dim
        slice(mid - margin, mid + margin),  # middle of dim
        slice(-chunk_size_1d, None))   # end of dim
        for mid in midpoints
    ]
    for i, sl in enumerate(slice_per_dim):
        if sl[0].stop > sl[1].start:  # slices overlap — just make that dim one chunk
            slice_per_dim[i] = (slice(None),)
    slices = list(itertools.product(*slice_per_dim))
    return [data[sl] for sl in slices]
```

This can be extended for RGB by reducing the sample size by 1/3, then sampling each channel independently, then concatenating the lists.

What do you think?

**Comment by Czaki** [comment]
Hm. I disagree that your code is more readable. I only get the existing code and add a special case for the plane to avoid sampling the same plane three times. 

Also, your code will fail current tests (as it will sample whole 1d data). 

I could try to sit longer on this to try to solve the problem instead of hiding it, but it may be out of the scope of 0.4.19 release manager.

**Comment by jni** [comment]
> Also, your code will fail current tests (as it will sample whole 1d data).

Huh? I had to modify code a bit to fix int/float bug and issue with asymmetric sample size in the middle, but:

```python
def sample_data_chunks(data, sample_size, chunks=3):
    ndim = data.ndim
    chunk_size_1d = int(sample_size ** (1/ndim) / chunks**ndim)
    left, right = math.floor(c := chunk_size_1d / 2), math.ceil(c)
    midpoints = tuple(s // 2 for s in data.shape)
    slice_per_dim = [
        (slice(None, chunk_size_1d),  # start of dim
        slice(mid - left, mid + right),  # middle of dim
        slice(-chunk_size_1d, None))   # end of dim
        for mid in midpoints
    ]
    for i, sl in enumerate(slice_per_dim):
        if sl[0].stop > sl[1].start:  # slices overlap — just make that dim one chunk
            slice_per_dim[i] = (slice(None),)
    slices = list(itertools.product(*slice_per_dim))
    return [data[sl] for sl in slices]
```

```
In [35]: sample_data_chunks(np.arange(10), 3)
Out[35]: [array([0]), array([5]), array([9])]
```

Anyway, my contention is not that my code is more readable in isolation. My contention is that compared to having `sample1d`, `sample2d` and `sample3d`, having a single n-dimensional function is better — and easier to maintain.

**Comment by jni** [comment]
> I could try to sit longer on this to try to solve the problem instead of hiding it, but it may be out of the scope of 0.4.19 release manager.

I can try to have a crack at it this afternoon... 🙏

**Comment by Czaki** [comment]
> Anyway, my contention is not that my code is more readable in isolation. My contention is that compared to having `sample1d`, `sample2d` and `sample3d`, having a single n-dimensional function is better — and easier to maintain.

I do not want to maintain any of this solution. When we have a numpy array, we could simply scan it as it is in memory. 
When having a distributed array, we should adjust scanning to chunk size. Otherwise, we could always have either performance or precision problems relatively often.

**Comment by Czaki** [comment]
close in favour of #6216
